src/binding/c/abi_api.txt

```diff
@@ -60,25 +50,47 @@ MPI_Abi_get_info:
     PUSH_TYPE_INFO(MPI_DOUBLE_COMPLEX, "mpi_double_complex_supported");
 #undef PUSH_TYPE_INFO
     *info = info_ptr->handle;
+#else
+    *info = MPI_INFO_NULL;
+#endif
+}
 
-  fn_fail:
-    return MPI_SUCCESS;
+MPI_Abi_get_info:
+    info: INFO, direction=out, [info object]
+    .desc: Return an info object that provides information related to the ABI
+    .seealso: MPI_Abi_get_version, MPI_Abi_get_fortran_info
+    .skip: initcheck
+{
+    MPIR_Info *info_ptr;
+    mpi_errno = MPIR_Info_alloc(&info_ptr);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    char str[8];
+#define PUSH_TYPE_INFO(type, keyname) \
+    do { \
+        snprintf(str, sizeof(str), "%d", (int) sizeof(type)); \
+        MPIR_Info_push(info_ptr, keyname, str); \
+    } while (0)
+    PUSH_TYPE_INFO(MPI_Aint, "mpi_aint_size");
+    PUSH_TYPE_INFO(MPI_Count, "mpi_count_size");
+    PUSH_TYPE_INFO(MPI_Offset, "mpi_offset_size");
+#undef PUSH_TYPE_INFO
+    *info = info_ptr->handle;
```

Comment:
We need rework this now that we separated the external MPI types from internal types. I'll push on to this PR.

The idea is for abi to never `HAVE_FORTRAN_BINDING` and rely on internal type table for type mappings.

Reply:
I know. As I stated in the description, this is a baseline PR to the get the "GET" part of the API working, such that I can implement some basic testing with mpi4py. 

> The idea is for abi to never `HAVE_FORTRAN_BINDING`

But then, if MPICH is built with NO Fortran, and the Fortran ABI is never set, what would get the output of `MPI_Abi_get_fortran_info()` ?

Reply:
The Fortran bindings should call `MPI_Abi_set_fortran_info()` to populate the type tables. 

> But then, if MPICH is built with NO Fortran, and the Fortran ABI is never set, what would get the output of MPI_Abi_get_fortran_info() ?

I would like to return `MPI_INFO_NULL` ;) In general, users (you) need handle the case when certain info keys are missing. None of the keys are required, are they?

Reply:
OK, OK... I'm just double-checking to see if we are in the same page... 

So, if `MPI_Abi_set_fortran_info()` is never called, then if `MPI_Abi_get_fortran_info()` is called, it will return: 
a) an info object populated with keys if MPICH was built **with** Fortran. 
b) `MPI_INFO_NULL` if MPICH was built **without** Fortran.

Is that correct?

Reply:
> Sidetrack, what is a "boolean" info key? Should we return the string "`.TRUE.`" and "`.FALSE.`"? That seems crazy. My instinct is to expect "yes" or "no"

Tagging @jeffhammond

Reply:
> b) `MPI_INFO_NULL` if MPICH was built **without** Fortran.

Nitpicking but `MPI_INFO_NULL` is typically special input value. In the case where there are no info keys associated with an output info object, a new object is still allocated. The new object just doesn't have any kv pairs. In other words the output of `MPI_INFO_GET_NKEYS` is 0.

Reply:
OK, but let me insist with a minor nit: there is a small semantic difference between returning an empty info (non-null but with no keyvals) from `MPI_INFO_NULL`. IMHO, the lack of a default Fortran ABI (that is, when MPICH is built with no Fortran) should be signaled with `MPI_INFO_NULL`.

> Sidetrack, what is a "boolean" info key? 
Do you mean for the various `mpi_xxx_supported` keys? I think`0` and `1`, I do not remember if the standard mandates the value in info for "boolean" things.

Reply:
Sorry for being so pedantic, but this is the kind of details that later on library developers strugle with different and have to resort to `#ifdef MPICH/OPEN_MPI`. We should try to avoid such workarounds by having a clear specification that everyone would implement right.

Reply:
> > Nitpicking but `MPI_INFO_NULL` is typically a special input value.
> 
> Please read the current text of the 5.0 standard. AFAICT, if MPICH knows nothing about Fortran, then `MPI_Abi_get_fortran_info()` must return `MPI_INFO_NULL`.

I stand corrected.

Reply:
Does MPI not already define the form of Boolean info key values?

Reply:
> Does MPI not already define the form of Boolean info key values?

There are many examples in the text. The defined values are "true" and "false".
```
"no_locks" (boolean, default: "false")
```

Reply:
It is easy to handing *input* info hints. Implementation can accept multiple ways of expressing booleans. I believe MPICH accepts `true/false`, `yes/no`, `1/0`. But here we are also sending info hints as output. Users would need standard way to parse the info hints.

Reply:
> > Does MPI not already define the form of Boolean info key values?
> 
> 
> 
> There are many examples in the text. The defined values are "true" and "false".
> 
> ```
> 
> "no_locks" (boolean, default: "false")
> 
> ```

Do I need to say it again or not? I guess more is better if it's ambiguous.

Reply:
Indeed there is normative text in the "info" chapter that says: Valid values for a boolean must include the strings "true" and "false" (all lowercase). That dispenses the ambiguity.

> Do I need to say it again or not? I guess more is better if it's ambiguous.
👍

Reply:
Sorry for my mistake. I was aware that MPI mandates true/false lowercase. I just kept the use of 1/0 that was already there without thinking twice. I'll not be hable to push a fix for about one week, but feel free to keep pushing to my branch, or take things over in a new PR and a new branch in your own repository fork.